### PR TITLE
cudnn: adjust version dependency on CUDA

### PR DESCRIPTION
**Pull request by melven**
According to the documentation, newer cudnn versions marked as "cuda
11.6" are compatible with all CUDA 11 versions.

See https://docs.nvidia.com/deeplearning/cudnn/support-matrix/index.html and specifically the remark https://docs.nvidia.com/deeplearning/cudnn/support-matrix/index.html#fntarg_2 .

This is the documentation for cudnn 8.4.1 but it's the same for 8.4.0 (https://docs.nvidia.com/deeplearning/cudnn/archives/cudnn-841/support-matrix/index.html).

### CODE REVIEW

File: var/spack/repos/builtin/packages/cudnn/package.py [hunk 1]
```diff
@@ -234,6 +234,9 @@ class Cudnn(Package):
             version(long_ver, sha256=pkg)
             # Add constraints matching CUDA version to cuDNN version
             cuda_req = 'cuda@{0}.0:{0}'.format(cuda_ver)
+            if cuda_ver == '11.6':
+                # cudnn with cuda 11.6 is upward compatible to newer cuda 11, see: https://docs.nvidia.com/deeplearning/cudnn/support-matrix/index.html#fntarg_2
+                cuda_req = 'cuda@11.0:11.999'
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cudnn/package.py]: Seems like this is true for all 11.X releases, not just 11.6. I have a proposed change but it's hard to explain. Do you mind if I push to your PR branch, or should I create a new PR?

Comment (melven) [reply var/spack/repos/builtin/packages/cudnn/package.py]: You can push here.

The documentation is tricky, 8.3.3 lists CUDA 11.5 which also means 11.X according to the docs... 

Some 8.2 versions support CUDA 11.0-11.4, not sure if the docs didn't get updated or if they actually don't work with 11.5+...

So, the CUDA version number in the cudnn version seems to give just a rough guess...

I can confirm that 8.4.1 works with 11.7 though in my setup.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/cudnn/package.py]: Okay, see if the latest commit still works for you. I also applied this to CUDA 12+ when they come out, we'll see if cuDNN continues this policy going forward.

Comment (melven) [reply var/spack/repos/builtin/packages/cudnn/package.py]: > Okay, see if the latest commit still works for you. I also applied this to CUDA 12+ when they come out, we'll see if cuDNN continues this policy going forward.

Does not work here - for any spec I obtain the following error now:
```console
$ spack spec ninja
==> Error: 11 does not satisfy 10.2
```

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/cudnn/package.py]: Hmm, why would ninja depend on cudnn? Also, what is the output of `spack debug report`?

Comment (melven) [reply var/spack/repos/builtin/packages/cudnn/package.py]: I get this error for any package somehow, see below:
```console
zoel_ml@fe-store01:~/spack_tmp (cudnn_cuda11_6)> spack debug report
* **Spack:** 0.19.0.dev0 (d98f40e2b735480b438db84aa2af5b50817c32d8)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-skylake_avx512
* **Concretizer:** clingo
zoel_ml@fe-store01:~/spack_tmp (cudnn_cuda11_6)> spack spec cudnn
==> Error: 11 does not satisfy 10.2
zoel_ml@fe-store01:~/spack_tmp (cudnn_cuda11_6)> spack spec cuda
==> Error: 11 does not satisfy 10.2
zoel_ml@fe-store01:~/spack_tmp (cudnn_cuda11_6)> spack spec ninja
==> Error: 11 does not satisfy 10.2
zoel_ml@fe-store01:~/spack_tmp (cudnn_cuda11_6)> git checkout HEAD~1
[...]
zoel_ml@fe-store01:~/spack_tmp ((6bc1bfa193...))> spack debug report
* **Spack:** 0.19.0.dev0 (6bc1bfa1935773a44ebffe4e1713bbbea998b0f4)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-skylake_avx512
* **Concretizer:** clingo
zoel_ml@fe-store01:~/spack_tmp ((6bc1bfa193...))> spack spec cudnn
Input spec
--------------------------------
cudnn

Concretized
--------------------------------
cudnn@8.4.0.27-11.6%gcc@12.1.0 arch=linux-ubuntu20.04-skylake_avx512
    ^cuda@11.7.0%gcc@12.1.0~allow-unsupported-compilers~dev arch=linux-ubuntu20.04-skylake_avx512
        ^libxml2@2.9.13%gcc@12.1.0~python arch=linux-ubuntu20.04-skylake_avx512
            ^libiconv@1.16%gcc@12.1.0 libs=shared,static arch=linux-ubuntu20.04-skylake_avx512
            ^pkgconf@1.8.0%gcc@12.1.0 arch=linux-ubuntu20.04-skylake_avx512
            ^xz@5.2.5%gcc@12.1.0~pic libs=shared,static arch=linux-ubuntu20.04-skylake_avx512
            ^zlib@1.2.12%gcc@12.1.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-skylake_avx512
```

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/cudnn/package.py]: Fixed, try again?

Comment (melven) [reply var/spack/repos/builtin/packages/cudnn/package.py]: Works fine now (sorry for the late reply...):

I tried to concretize various versions and got the expected results:
```console
> for CUDNN_VER in 8.4.0.27-11.6 8.4.0.27-10.2 8.3.3.40-11.5 8.3.3.40-10.2; do for CUDA_VER in 10.2 11.0 11.1 11.2 11.3 11.4 11.5 11.6 11.7; do cmd="spack spec cudnn@$CUDNN_VER ^cuda@$CUDA_VER"; echo $cmd; $cmd > /dev/null && echo "==> Success"; done; done
spack spec cudnn@8.4.0.27-11.6 ^cuda@10.2
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.7.0'
spack spec cudnn@8.4.0.27-11.6 ^cuda@11.0
==> Success
spack spec cudnn@8.4.0.27-11.6 ^cuda@11.1
==> Success
spack spec cudnn@8.4.0.27-11.6 ^cuda@11.2
==> Success
spack spec cudnn@8.4.0.27-11.6 ^cuda@11.3
==> Success
spack spec cudnn@8.4.0.27-11.6 ^cuda@11.4
==> Success
spack spec cudnn@8.4.0.27-11.6 ^cuda@11.5
==> Success
spack spec cudnn@8.4.0.27-11.6 ^cuda@11.6
==> Success
spack spec cudnn@8.4.0.27-11.6 ^cuda@11.7
==> Success
spack spec cudnn@8.4.0.27-10.2 ^cuda@10.2
==> Success
spack spec cudnn@8.4.0.27-10.2 ^cuda@11.0
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.0.3'
spack spec cudnn@8.4.0.27-10.2 ^cuda@11.1
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.1.1'
spack spec cudnn@8.4.0.27-10.2 ^cuda@11.2
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.2.2'
spack spec cudnn@8.4.0.27-10.2 ^cuda@11.3
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.3.1'
spack spec cudnn@8.4.0.27-10.2 ^cuda@11.4
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.4.4'
spack spec cudnn@8.4.0.27-10.2 ^cuda@11.5
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.5.2'
spack spec cudnn@8.4.0.27-10.2 ^cuda@11.6
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.6.2'
spack spec cudnn@8.4.0.27-10.2 ^cuda@11.7
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.7.0'
spack spec cudnn@8.3.3.40-11.5 ^cuda@10.2
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.7.0'
spack spec cudnn@8.3.3.40-11.5 ^cuda@11.0
==> Success
spack spec cudnn@8.3.3.40-11.5 ^cuda@11.1
==> Success
spack spec cudnn@8.3.3.40-11.5 ^cuda@11.2
==> Success
spack spec cudnn@8.3.3.40-11.5 ^cuda@11.3
==> Success
spack spec cudnn@8.3.3.40-11.5 ^cuda@11.4
==> Success
spack spec cudnn@8.3.3.40-11.5 ^cuda@11.5
==> Success
spack spec cudnn@8.3.3.40-11.5 ^cuda@11.6
==> Success
spack spec cudnn@8.3.3.40-11.5 ^cuda@11.7
==> Success
spack spec cudnn@8.3.3.40-10.2 ^cuda@10.2
==> Success
spack spec cudnn@8.3.3.40-10.2 ^cuda@11.0
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.0.3'
spack spec cudnn@8.3.3.40-10.2 ^cuda@11.1
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.1.1'
spack spec cudnn@8.3.3.40-10.2 ^cuda@11.2
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.2.2'
spack spec cudnn@8.3.3.40-10.2 ^cuda@11.3
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.3.1'
spack spec cudnn@8.3.3.40-10.2 ^cuda@11.4
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.4.4'
spack spec cudnn@8.3.3.40-10.2 ^cuda@11.5
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.5.2'
spack spec cudnn@8.3.3.40-10.2 ^cuda@11.6
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.6.2'
spack spec cudnn@8.3.3.40-10.2 ^cuda@11.7
==> Error: No version for 'cuda' satisfies '@10.2.89' and '@11.7.0'
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**